tests_metricflow/dataflow/optimizer/test_predicate_pushdown_optimizer.py

```diff
@@ -0,0 +1,150 @@
+from __future__ import annotations
+
+from datetime import datetime
+
+import pytest
+from metricflow_semantics.filters.time_constraint import TimeRangeConstraint
+from metricflow_semantics.specs.spec_classes import WhereFilterSpec
+from metricflow_semantics.sql.sql_bind_parameters import SqlBindParameters
+
+from metricflow.dataflow.optimizer.predicate_pushdown_optimizer import PredicatePushdownBranchStateTracker
+from metricflow.plan_conversion.node_processor import PredicateInputType, PredicatePushdownState
+
+
+@pytest.fixture
+def fully_enabled_pushdown_state() -> PredicatePushdownState:
+    """Provides a valid configuration with all predicate properties set and pushdown fully enabled."""
+    params = PredicatePushdownState(time_range_constraint=TimeRangeConstraint.all_time(), where_filter_specs=tuple())
+    return params
+
+
+@pytest.fixture
+def branch_state_tracker(fully_enabled_pushdown_state: PredicatePushdownState) -> PredicatePushdownBranchStateTracker:
+    """Provides a branch state tracker for direct testing of update mechanics."""
+    return PredicatePushdownBranchStateTracker(initial_state=fully_enabled_pushdown_state)
+
+
+def test_time_range_pushdown_enabled_states(fully_enabled_pushdown_state: PredicatePushdownState) -> None:
+    """Tests pushdown enabled check for time range pushdown operations."""
+    time_range_only_state = PredicatePushdownState(
+        time_range_constraint=TimeRangeConstraint.all_time(),
+        pushdown_enabled_types=frozenset([PredicateInputType.TIME_RANGE_CONSTRAINT]),
+        where_filter_specs=tuple(),
+    )
+
+    enabled_states = {
+        "fully enabled": fully_enabled_pushdown_state.has_time_range_constraint_to_push_down,
+        "enabled for time range only": time_range_only_state.has_time_range_constraint_to_push_down,
+    }
+
+    assert all(list(enabled_states.values())), (
+        "Expected pushdown to be enabled for pushdown state with time range constraint and global pushdown enabled, "
+        "but some states returned False for has_time_range_constraint_to_push_down.\n"
+        f"Pushdown enabled states: {enabled_states}\n"
+        f"Fully enabled state: {fully_enabled_pushdown_state}\n"
+        f"Time range only state: {time_range_only_state}"
+    )
+
+
+def test_invalid_disabled_pushdown_state() -> None:
+    """Tests checks for invalid param configuration on disabled pushdown parameters."""
+    with pytest.raises(AssertionError, match="Disabled pushdown state objects cannot have properties set"):
+        PredicatePushdownState(
+            time_range_constraint=TimeRangeConstraint.all_time(),
+            pushdown_enabled_types=frozenset(),
+            where_filter_specs=tuple(),
+        )
```

Comment:
Apart from the branch_state_tracker method, everything above this line existed in the previous file. Graphite shows this nicely. GitHub, as always, is deficient.